train/compute/python/tools/trace_link.py

```diff
@@ -80,8 +82,127 @@ def find_closest_segment(segs, target_length):
+        return op_dict["dur"]
+    else:
+        return 0
+
+
+def get_timestamp_field(op_dict):
+    if "ts" in op_dict:
+        return op_dict["ts"]
+    else:
+        return 0
+
+
+def find_closest_op(ops, ts):
+    """
+    Finds the operation that is closest in duration to a given timestamp.
+    """
+    closest_start = 0
+    closest_op = {}
+    for op in ops:
+        if (
+            has_category_field(op)
+            and get_timestamp_field(op) < ts
+            and (get_timestamp_field(op) + get_duration_field(op)) > ts
+            and get_timestamp_field(op) > closest_start
+        ):
+
+            closest_start = get_timestamp_field(op)
+            closest_op = op
+    return closest_op
+
+
+def find_parent_cpu_op(
+    kineto_gpu_op,
+    kineto_et_ops,
+    kineto_ac2g_s_ops,
+    kineto_ac2g_f_ops,
+    kineto_cpu_launcher_ops,
+):
+    """
+    Find the parent CPU operation for a given GPU operation based on
+    the closest start time and duration that covers the GPU operation.
+    """
+    ts = -1
+
+    # Find a CPU launch op which has the same external id with the GPU op.
+    if kineto_gpu_op["args"]["External id"] in kineto_cpu_launcher_ops.keys():
+        ts = get_timestamp_field(
+            kineto_cpu_launcher_ops[kineto_gpu_op["args"]["External id"]]
+        ) + get_duration_field(
+            kineto_cpu_launcher_ops[kineto_gpu_op["args"]["External id"]]
+        )
+    # Find an arrow (CPU->GPU) start point which has the same external id with the GPU op.
+    elif kineto_gpu_op["args"]["External id"] in kineto_ac2g_s_ops.keys():
+        ts = get_timestamp_field(
+            kineto_ac2g_s_ops[kineto_gpu_op["args"]["External id"]]
+        )
+    # Find an arrow (CPU->GPU) final destination point which has the same external id with the GPU op.
+    elif kineto_gpu_op["args"]["External id"] in kineto_ac2g_f_ops.keys():
+        ts = get_timestamp_field(
+            kineto_ac2g_f_ops[kineto_gpu_op["args"]["External id"]]
+        )
+    assert ts != -1
+
+    # Find a parent cpu launch op based on two conditions.
+    # Condition 1: Parent op should be the op has the closest start time to the gpu op.
+    # Condition 2: Parent op should be long enough to cover the gpu op.
+    kineto_gpu_op["ts"] = ts
+    parent_cpu_op = find_closest_op(kineto_et_ops, ts)
+
+    if not parent_cpu_op:
+        print(
+            "Warning! the parent cpu_op for the following gpu_op is not found"
+            + "and hence it is discarded. gpu_op name: "
+            + str(kineto_gpu_op["name"])
+            + ", ts: "
+            + str(get_timestamp_field(kineto_gpu_op))
+            + ", external_id: "
+            + str(kineto_gpu_op["args"]["External id"])
+        )
+
+    # TODO: In some cases, condition 2 is not met. Seems to be a problem in input from older Kineto.
+    # assert parent_cpu_op and get_timestamp_field(parent_cpu_op) + get_duration_field(parent_cpu_op) > get_timestamp_field(kineto_gpu_op)
```

Comment:
@JoongunPark Could you please remove commented-out codes?

Reply:
@TaekyungHeo Yes. I have removed.